airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJdbcSource.java

```diff
@@ -175,21 +181,60 @@ public AirbyteCatalog discover(JsonNode config) throws Exception {
       }
 
       final String fieldNames = selectedDatabaseFields.stream().map(Field::getName).collect(Collectors.joining(", "));
-      final Stream<AirbyteMessage> stream = database.query(
-          ctx -> ctx.fetchStream(String.format("SELECT %s FROM %s", fieldNames, table.getName()))
-              .map(r -> new AirbyteMessage()
-                  .withType(Type.RECORD)
-                  .withRecord(new AirbyteRecordMessage()
-                      .withStream(airbyteStream.getStream().getName())
-                      .withEmittedAt(now.toEpochMilli())
-                      .withData(Jsons.deserialize(r.formatJSON(DB_JSON_FORMAT))))));
+
+      final Stream<AirbyteMessage> stream;
+      if (airbyteStream.getSyncMode() == SyncMode.INCREMENTAL) {
+        final JdbcStateManager stateManager = new JdbcStateManager(Jsons.object(state, JdbcState.class));
+        final String cursorField = IncrementalUtils.getCursorField(airbyteStream);
+        final JsonSchemaPrimitive cursorType = IncrementalUtils.getCursorType(airbyteStream, cursorField);
+        final Optional<String> initialCursorOptional = stateManager.getOriginalCursor(streamName);
+
+        final Stream<AirbyteMessage> internalMessageStream;
+        if (initialCursorOptional.isPresent()) {
+          internalMessageStream =
+              getMessageStream(executeIncrementalQuery(database, fieldNames, table.getName(), cursorField, initialCursorOptional.get()), streamName,
+                  now.toEpochMilli());
+        } else {
+          internalMessageStream = getMessageStream(executeFullRefreshQuery(database, fieldNames, table.getName()), streamName, now.toEpochMilli());
+        }
+
+        final StateDecoratingIterator stateDecoratingIterator = new StateDecoratingIterator(
+            internalMessageStream,
+            stateManager,
+            streamName,
+            cursorField,
+            initialCursorOptional.orElse(null),
+            cursorType);
+
+        stream = MoreStreams.toStream(stateDecoratingIterator);
+      } else {
+        stream = getMessageStream(executeFullRefreshQuery(database, fieldNames, table.getName()), streamName, now.toEpochMilli());
+      }
 
       resultStream = Stream.concat(resultStream, stream);
     }
 
     return resultStream.onClose(() -> Exceptions.toRuntime(database::close));
   }
 
+  private static Stream<Record> executeFullRefreshQuery(Database database, String fieldNames, String tableName) throws SQLException {
+    return database.query(ctx -> ctx.fetchStream(String.format("SELECT %s FROM %s", fieldNames, tableName)));
+  }
+
+  private static Stream<Record> executeIncrementalQuery(Database database, String fieldNames, String tableName, String cursorField, String cursor)
+      throws SQLException {
+    return database.query(ctx -> ctx.fetchStream(String.format("SELECT %s FROM %s WHERE %s > %s", fieldNames, tableName, cursorField, cursor)));
```

Comment:
I need to rethink this comparison here. I think for anything but number types this syntax won't work. need to figure how to cast to the database type and then do the comparison. same when calculating `maxCursor` below.